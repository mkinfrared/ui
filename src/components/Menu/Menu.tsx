import { useEffect, useLayoutEffect, useRef } from "react";
import { createPortal } from "react-dom";

import { useOutsideClick } from "hooks";
import { classNames } from "utils";

import css from "./Menu.module.scss";
import { MenuProps } from "./Menu.type";

const Menu = ({ className, children, comboboxRef, id, onClose }: MenuProps) => {
  const { body } = document;
  const ref = useRef<HTMLUListElement>(null);

  useOutsideClick([ref, comboboxRef], onClose);

  const computeYAxis = (combobox: HTMLElement, dropdownMenu: HTMLElement) => {
    const { top, bottom } = combobox.getBoundingClientRect();
    const { height: comboboxHeight } = dropdownMenu.getBoundingClientRect();
    const nextBottomY = bottom + comboboxHeight;
    const nextTopY = top - comboboxHeight;
    const visualOffset = Math.ceil(window.visualViewport?.offsetTop ?? 0);
    const visualHeight = window.visualViewport?.height ?? Infinity;
    const windowHeight = Math.min(window.innerHeight, visualHeight);

    // if dropdown bottom edge doesn't show up outside the viewport
    // place it at the bottom of combobox
    if (nextBottomY < windowHeight) {
      dropdownMenu.style.setProperty("--top", `${bottom}px`);

      dropdownMenu.style.setProperty("--bottom", "auto");

      dropdownMenu.style.setProperty("--translate-y", `${visualOffset}px`);

      // else if dropdown top edge doesn't show up outside the viewport
      // place it at the top of combobox
    } else if (nextTopY >= 0) {
      dropdownMenu.style.setProperty(
        "--bottom",
        `${window.innerHeight - top}px`,
      );

      dropdownMenu.style.setProperty("--top", "auto");

      dropdownMenu.style.setProperty("--translate-y", `${visualOffset}px`);
      // otherwise place it at the bottom of combobox
    } else {
      dropdownMenu.style.setProperty("--top", `${bottom}px`);

      dropdownMenu.style.setProperty("--bottom", "auto");

      dropdownMenu.style.setProperty("--translate-y", `${visualOffset}px`);
    }
  };

  const computeXAxis = (combobox: HTMLElement, dropdownMenu: HTMLElement) => {
    const { left, right } = combobox.getBoundingClientRect();
    const { width: comboboxWidth } = dropdownMenu.getBoundingClientRect();
    const nextRightX = left + comboboxWidth;
    const nextLeftX = right - comboboxWidth;

    // if dropdown right edge doesn't show up outside the viewport
    // place it at the left of combobox
    if (nextRightX < window.innerWidth) {
      dropdownMenu.style.setProperty("--left", `${left}px`);

      // else if dropdown left edge doesn't show up outside the viewport
      // place it at the right of combobox
    } else if (nextLeftX >= 0) {
      dropdownMenu.style.setProperty("--left", `${nextLeftX}px`);
      // otherwise place it at the left of viewport and set the width of viewport
    } else {
      dropdownMenu.style.setProperty("--left", "0px");

      dropdownMenu.style.setProperty(
        "--combobox-width",
        `${window.innerWidth}px`,
      );
    }
  };

  const setPosition = (comboboxElement: HTMLElement) => {
    const { current } = ref;

    if (!current) {
      return;
    }

    const { width } = comboboxElement.getBoundingClientRect();

    current.style.setProperty("--combobox-width", `${width}px`);

    computeYAxis(comboboxElement, current);

    computeXAxis(comboboxElement, current);
  };

  useLayoutEffect(() => {
    if (comboboxRef.current) {
      setPosition(comboboxRef.current);
    }
  }, []);

  useEffect(() => {
    const { height, overflow } = document.body.style;

    document.body.style.height = "100vh";

    document.body.style.overflow = "hidden";

    const handleResize = () => {
      if (comboboxRef.current) {
        setPosition(comboboxRef.current);
      }
    };

    const handleKeyDown = (event: KeyboardEvent) => {
      const { key } = event;

      if (key === "Escape") {
        onClose();
      }
    };

    window.addEventListener("resize", handleResize);

    window.visualViewport?.addEventListener("scroll", handleResize);

    window.visualViewport?.addEventListener("resize", handleResize);

    window.addEventListener("scroll", (event) => {
      event.preventDefault();

      handleResize();
    });

    window.addEventListener("keydown", handleKeyDown);

    return () => {
      window.removeEventListener("resize", handleResize);

      window.visualViewport?.removeEventListener("scroll", handleResize);

      window.visualViewport?.removeEventListener("resize", handleResize);

      window.removeEventListener("keydown", handleKeyDown);

      document.body.style.height = height;

      document.body.style.overflow = overflow;
    };
  }, []);

  return createPortal(
    <ul
      className={classNames(css.Menu, className)}
      data-testid="Menu"
      id={id}
      ref={ref}
      role="listbox"
      onKeyDown={() => null}
    >
      {children}
    </ul>,
    body,
  );
};

export default Menu;
