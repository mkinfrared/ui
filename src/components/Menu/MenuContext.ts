import { ReactNode, createContext } from "react";

type MenuContextType<T> = {
  onClick?: (value: T | "", children: ReactNode) => void;
  currentValue?: T | "";
};

const MenuContext = createContext<MenuContextType<any>>({});

export { MenuContext };
