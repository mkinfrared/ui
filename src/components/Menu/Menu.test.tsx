import { render } from "@testing-library/react";

import Menu from "./Menu";

describe("<Menu />", () => {
  const comboboxRef = { current: null };
  const id = "marklar";
  const onClose = jest.fn();

  const Component = (
    <Menu comboboxRef={comboboxRef} id={id} onClose={onClose}>
      <li>Item-1</li>
      <li>Item-2</li>
      <li>Item-3</li>
      <li>Item-4</li>
      <li>Item-5</li>
    </Menu>
  );

  it("should be defined", () => {
    expect(Menu).toBeDefined();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("Menu");

    expect(element).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("Menu");

    expect(element).toMatchSnapshot();
  });
});
