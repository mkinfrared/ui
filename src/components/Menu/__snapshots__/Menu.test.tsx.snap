// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should match the snapshot 1`] = `
<ul
  class="Menu"
  data-testid="Menu"
  id="marklar"
  role="listbox"
>
  <li>
    Item-1
  </li>
  <li>
    Item-2
  </li>
  <li>
    Item-3
  </li>
  <li>
    Item-4
  </li>
  <li>
    Item-5
  </li>
</ul>
`;
